contracts/semi-bridge/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use ink_lang as ink;
+
+#[ink::contract(env = pink_extension::PinkEnvironment)]
+mod semi_bridge {
+    use alloc::{string::String, string::ToString, vec::Vec};
+    use index::prelude::*;
+    use index::utils::ToArray;
+    use ink_storage::traits::{PackedLayout, SpreadLayout};
+    use pink_web3::ethabi::{Bytes, Uint};
+    use pink_web3::keys::pink::KeyPair;
+    use pink_web3::signing::Key;
+    use primitive_types::{H160, H256, U256};
+    use scale::{Decode, Encode};
+
+    #[ink(storage)]
+    pub struct SemiBridge {
+        owner: AccountId,
+        key: [u8; 32],
+        config: Option<Config>,
+    }
+
+    #[derive(Encode, Decode, Debug, PackedLayout, SpreadLayout)]
+    #[cfg_attr(
+        feature = "std",
+        derive(scale_info::TypeInfo, ink_storage::traits::StorageLayout)
+    )]
+    struct Config {
+        rpc: String,
+        bridge_address: [u8; 20],
+    }
+
+    #[derive(Encode, Decode, Debug, PartialEq, Eq)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        BadOrigin,
+        NotConfigurated,
+        KeyRetired,
+        KeyNotRetiredYet,
+        UpstreamFailed,
+        BadAbi,
+        FailedToGetStorage,
+        FailedToDecodeStorage,
+        FailedToEstimateGas,
+        FailedToCreateExecutor,
+    }
+
+    type Result<T> = core::result::Result<T, Error>;
+
+    impl SemiBridge {
+        #[ink(constructor)]
+        pub fn default() -> Self {
+            Self {
+                owner: Self::env().caller(),
+                config: None,
+                key: Self::key_pair().private_key(),
```

Comment:
The key will be allocated by a dedicated worker account manager and assigned to the bridge executor during plan execution by the global executor. We could consider adding a parameter `key` in the `transfer` function or add another `set_key` function which is being called when executing and call `clear_key` to detach the key with the bridge executor.

Reply:
certainly, there is already a `set_account` interface that takes a private key as an argument in line 76

is the `set_key` function you said the same as `set_account`?

and is the `key` parameter you suggested to be added to the `transfer` function a private key?

if so, the interfaces will look like this:

```rust
fn set_key(&mut self, private_key: [u8; 32]);
fn transfer(&self, private_key: [u8;32], ....);
```

So, if we decide to add a private key parameter to the `transfer()` function, then what is the point of `set_key()`? because in the new `transfer()` function we are definitely not using `self.private_key` to sign tx.

Reply:
I mean we can choose either of them. And yes `set_key` same as `set_account`.